### PR TITLE
Start of exhaustively documenting the build file.

### PR DESCRIPTION
**Pull request by vaikas**
While learning about, couldn't really find one place to understand the build file, fields, etc. so figured I'd take a crack at creating one. Not comprehensive, but a start to see if keeping on going makes sense.

### CODE REVIEW

**Comment by vaikas** [comment]
There are bunch of TODO(vaikas) that I need some help with, so just jotted them down while I was writing. I'll keep iterating.

File: docs/BUILD-FILE.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
+# melange build file
+
+This documents the melange build file structure, fields, when, and why to use
+various fields.
```
[the remark this answers is not in the document]
Comment (vaikas) [reply docs/BUILD-FILE.md:4]: are you seriously suggesting on not wrapping at 80???

File: docs/BUILD-FILE.md [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+   Package metadata about this package, name, version, etc.
+ * environment
+
+    Specification for the packages build environment
+ * pipeline
+
+    Ordered list of pipelines that produce this package
+
+And the following optional sections:
```
[the remark this answers is not in the document]
Comment (vaikas) [reply docs/BUILD-FILE.md:21]: good idea, I'm not so good with markdown.

**Comment by vaikas** [comment]
Addressed feeback, and I added some great verbiage from @erikaheidi here.

**Comment by vaikas** [comment]
Addressed feeback, and I added some great verbiage from @erikaheidi here.

File: docs/BUILD-FILE.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (stormqueen1990) [reply docs/BUILD-FILE.md]: @jonjohnsonjr my bad for disagreeing but as someone who spends a lot of time comparing markdown files I appreciate the wrapping 😬

**Review by imjasonh — APPROVED** [review]

**Pull request merged**